packages/@lwc/engine-dom/src/apis/create-element.ts

```diff
@@ -98,21 +98,28 @@ if (process.env.NODE_ENV !== 'production') {
 }
 
 /**
- * Properties defined on the component class, excluding those inherited from `LightningElement`.
+ * Gets the public properties of a component class. If the `__lwc_public_property_types__` property
+ * is defined, it will be used as the source of truth for the property types. Otherwise, all of the
+ * properties defined on the component class are used, excluding those inherited from `LightningElement`.
+ *
+ * IMPORTANT: If the fallback is used, then _all_ component properties are returned, rather than
+ * just the public properties.
  */
-// TODO [#4292]: Restrict this to only @api props
-type ComponentClassProperties<T> = Omit<T, keyof LightningElement>;
+type ComponentClassProperties<T> = T extends {
+    __lwc_public_property_types__?: infer Props extends object;
```

Comment:
```suggestion
    readonly __lwc_public_property_types__?: infer Props extends object;
```
Readonly props don't extend mutable props, but mutable props can extend readonly props, so let's use the `readonly` modifier so that we can catch more.